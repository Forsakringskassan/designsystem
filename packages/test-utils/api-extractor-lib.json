{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
