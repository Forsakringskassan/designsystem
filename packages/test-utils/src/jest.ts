/* imported for side-effects only */
import "./matchers";

export * from "./lib";
