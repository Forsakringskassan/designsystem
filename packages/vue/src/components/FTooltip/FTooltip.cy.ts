import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import { defineComponent, useTemplateRef } from "vue";
import { FTooltipPageObject } from "../../cypress";
import { FLabel } from "../FLabel";
import { FTextField } from "../FTextField";
import FTooltip from "./FTooltip.vue";

describe("FTooltip", () => {
    it("Should have a page object that can access any necessary elements", () => {
        cy.mount(FTooltip, {
            slots: {
                header: `Rubrik`,
                body: /* HTML */ ` <p>Lorem ipsum</p> `,
            },
            props: {
                screenReaderText: "Screen reader text",
                headerTag: "h3",
            },
        });
        const tooltip = new FTooltipPageObject(".tooltip");
        tooltip.iButton().click();
        tooltip.header().should("contain", "Rubrik");
        tooltip.body().should("contain", "Lorem ipsum");
        tooltip.closeButton().should("have.trimmedText", "Stäng");
        tooltip.closeButton().click();
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip("should appear visually correct with label and tooltip", () => {
        const TestComponent = defineComponent({
            components: { FTooltip },
            template: /* HTML */ `
                <div v-for="(v, index) in variants">
                    <div ref="attach">
                        <label class="label"> Etikett </label>
                    </div>
                    <f-tooltip
                        :attach-to="attach?.[index]"
                        :id="v.id"
                        v-model="v.model"
                        screen-reader-text="Läs mer"
                    >
                        <template #header> Rubrik </template>
                        <template #body> Lorem ipsum dolor sit amet. </template>
                    </f-tooltip>
                </div>
            `,
            setup() {
                return { attach: useTemplateRef("attach") };
            },
            data() {
                return {
                    variants: [
                        { id: "collapsed", model: false },
                        { id: "expanded", model: true },
                    ],
                };
            },
        });
        cy.viewport(800, 600);
        cy.mount(TestComponent);
        cy.get("#collapsed .tooltip__body").should("not.be.visible");
        cy.get("#expanded .tooltip__body").should("be.visible");
        cy.toMatchScreenshot();
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip("should appear visually correct with heading and tooltip", () => {
        const TestComponent = defineComponent({
            components: { FTooltip },
            template: /* HTML */ `
                <div v-for="(v, index) in variants">
                    <div ref="attach">
                        <h3 class="label">Rubrik</h3>
                    </div>
                    <f-tooltip
                        :attach-to="attach?.[index]"
                        :id="v.id"
                        v-model="v.model"
                        screen-reader-text="Läs mer"
                    >
                        <template #header> Rubrik </template>
                        <template #body> Lorem ipsum dolor sit amet. </template>
                    </f-tooltip>
                </div>
            `,
            setup() {
                return { attach: useTemplateRef("attach") };
            },
            data() {
                return {
                    variants: [
                        { id: "collapsed", model: false },
                        { id: "expanded", model: true },
                    ],
                };
            },
        });
        cy.viewport(800, 600);
        cy.mount(TestComponent);
        cy.get("#collapsed .tooltip__body").should("not.be.visible");
        cy.get("#expanded .tooltip__body").should("be.visible");
        cy.toMatchScreenshot();
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip("should appear visually correct with input field and tooltip", () => {
        const TestComponent = defineComponent({
            components: { FTextField, FTooltip },
            template: /* HTML */ `
                <f-text-field v-for="v in variants">
                    <template #default> Etikett </template>
                    <template #tooltip>
                        <f-tooltip
                            :id="v.id"
                            v-model="v.model"
                            screen-reader-text="Läs mer"
                        >
                            <template #header> Rubrik </template>
                            <template #body>
                                Lorem ipsum dolor sit amet.
                            </template>
                        </f-tooltip>
                    </template>
                </f-text-field>
            `,
            data() {
                return {
                    variants: [
                        { id: "collapsed", model: false },
                        { id: "expanded", model: true },
                    ],
                };
            },
        });
        cy.viewport(800, 600);
        cy.mount(TestComponent);
        cy.get("#collapsed .tooltip__body").should("not.be.visible");
        cy.get("#expanded .tooltip__body").should("be.visible");
        cy.toMatchScreenshot();
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    describe.skip("density", () => {
        const DensityComponent = defineComponent({
            template: /* HTML */ `
                <density-wrapper>
                    <f-label>
                        <template #default> Tooltip </template>
                        <template #tooltip>
                            <f-tooltip screen-reader-text="Skärmläsartext">
                                <template #header> Rubrik </template>
                                <template #body> Brödtext </template>
                            </f-tooltip>
                        </template>
                    </f-label>
                </density-wrapper>
            `,
            components: {
                DensityWrapper,
                FLabel,
                FTooltip,
            },
        });

        it(`should be densified`, () => {
            cy.viewport(densityWrapperWidth, densityWrapperHeight);
            cy.mount(DensityComponent);
            const tooltip1 = new FTooltipPageObject(
                ".density-default .tooltip",
            );
            tooltip1.iButton().click({ multiple: true });
            const tooltip2 = new FTooltipPageObject(".density-dense .tooltip");
            tooltip2.iButton().click({ multiple: true });
            const tooltip3 = new FTooltipPageObject(
                ".density-densest .tooltip",
            );
            tooltip3.iButton().click({ multiple: true });
            cy.toMatchScreenshot();
        });
    });
});
