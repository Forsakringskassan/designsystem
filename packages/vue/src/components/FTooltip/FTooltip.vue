<template>
    <div class="tooltip">
        <div class="tooltip__container">
            <button
                class="tooltip__button"
                type="button"
                :aria-expanded="isOpen ? 'true' : 'false'"
                @click="onClickToggle"
            >
                <span class="icon-stack icon-stack--tooltip">
                    <f-icon name="circle"></f-icon>
                    <f-icon name="i"></f-icon>
                    <span class="sr-only">{{ screenReaderText }}</span>
                </span>
            </button>
            <f-expand>
                <div v-if="isOpen">
                    <div class="tooltip__content-wrapper" tabindex="-1" :aria-hidden="isOpen ? undefined : 'true'">
                        <span v-show="isOpen" class="tooltip__arrow"></span>
                        <div class="tooltip__content">
                            <component :is="headerTag" v-if="hasHeader" class="tooltip__header">
                                <!-- @slot Tooltip header content -->
                                <slot name="header"></slot>
                            </component>

                            <div class="tooltip__body">
                                <!-- @slot Tooltip body content-->
                                <slot name="body"></slot>
                            </div>
                        </div>
                        <i-flex float="right">
                            <i-flex-item shrink>
                                <button class="close-button" type="button" @click="onClickToggle">
                                    <span>{{ closeButtonText }}</span>
                                    <f-icon class="button__icon" name="close"></f-icon>
                                </button>
                            </i-flex-item>
                        </i-flex>
                    </div>
                </div>
            </f-expand>
        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { focus, TranslationService } from "@fkui/logic";
import { FExpand } from "../FExpand";
import { IFlex, IFlexItem } from "../../internal-components";
import { FIcon } from "../FIcon";
import { hasSlot } from "../../utils";

export default defineComponent({
    name: "FTooltip",
    components: { FExpand, FIcon, IFlex, IFlexItem },
    props: {
        /**
         * State (expanded or collapsed) of the tooltip. The value is `true` if the tooltip is expanded.
         *
         * @model
         */
        modelValue: {
            type: Boolean,
            required: false,
        },
        /**
         * Screen reader text for toggle button
         */
        screenReaderText: {
            type: String,
            required: true,
        },
        /**
         * Close button text
         */
        closeButtonText: {
            type: String,
            required: false,
            default: TranslationService.provider.translate("fkui.tooltip.close", "StÃ¤ng"),
        },
        /**
         * Element to render for the header element inside the tooltip.
         *
         * Must be set to one of:
         *
         * - `div` (default)
         * - `span`
         * - `h1`
         * - `h2`
         * - `h3`
         * - `h4`
         * - `h5`
         * - `h6`
         */
        headerTag: {
            default: "div",
            required: false,
            validator(value: string): boolean {
                return ["div", "span", "h1", "h2", "h3", "h4", "h5", "h6"].includes(value);
            },
        },
    },
    emits: ["update:modelValue", "toggle"],
    data() {
        return {
            isOpen: false,
        };
    },
    computed: {
        hasHeader(): boolean {
            return hasSlot(this, "header");
        },
    },
    watch: {
        modelValue: {
            immediate: true,
            handler(value: boolean) {
                this.isOpen = value;
            },
        },
    },
    mounted() {
        window.addEventListener("resize", () => {
            if (this.isOpen) {
                this.positionArrow();
            }
        });

        if (this.isOpen) {
            this.positionArrow();
        }
    },
    methods: {
        /**
         * Gets called when the user interacts with the toggle button
         *
         * @internal
         */
        onClickToggle(): void {
            this.isOpen = !this.isOpen;

            const value = this.isOpen;
            const event = { isOpen: this.isOpen };

            /**
             * v-model event.
             *
             * @param {boolean} value - Model value
             */
            this.$emit("update:modelValue", value);

            /**
             * Emitted when the state of the tooltip (collapsed/expanded) changes.
             *
             * @param {{ isOpen: boolean }} event - New state of tooltip.
             */
            this.$emit("toggle", event);

            if (!this.isOpen) {
                const button = this.$el.querySelector(".tooltip__button");
                focus(button);
            }
            this.$nextTick(() => {
                this.positionArrow();
            });
        },
        positionArrow(): void {
            const button: HTMLElement | null = this.$el.querySelector(".tooltip__button");
            const arrow: HTMLElement | null = this.$el.querySelector(".tooltip__arrow");
            const content: HTMLElement | null = this.$el.querySelector(".tooltip__content-wrapper");
            const borderSize = 2;

            if (button && arrow && content) {
                const buttonOffsetLeft: number = button.offsetLeft - content.offsetLeft;
                const relativeOffset = buttonOffsetLeft - borderSize + button.getBoundingClientRect().width / 2;
                arrow.style.left = `${relativeOffset}px`;
            }
        },
    },
});
</script>
