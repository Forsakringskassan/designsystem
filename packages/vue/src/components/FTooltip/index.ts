export { default as FTooltip } from "./FTooltip.vue";
