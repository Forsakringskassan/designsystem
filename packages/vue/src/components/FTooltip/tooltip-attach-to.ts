import { type InjectionKey, type ShallowRef } from "vue";

/**
 * @internal
 */
export const tooltipAttachTo = Symbol("tooltipAttachTo") as InjectionKey<
    Readonly<ShallowRef<HTMLElement | null>>
>;
