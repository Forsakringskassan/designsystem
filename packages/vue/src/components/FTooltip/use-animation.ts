import { type Ref, ref, computed, onMounted, watchEffect } from "vue";

let initialized = false;
const reducedMotion = ref(false);

/**
 * @internal
 */
export interface UseAnimation {
    /**
     * Set to `true` if the user has animations enabled (i.e.
     * `prefers-reduced-motion` is not set).
     */
    enabled: Readonly<Ref<boolean>>;

    animate(state: "expand" | "collapse"): void;
}

/**
 * @internal
 */
export function useAnimation(options: {
    duration?: number;
    easing?: string;
    element: Readonly<Ref<HTMLElement | null>>;
}): UseAnimation {
    const { duration = 250, easing = "ease-in", element: elementRef } = options;

    let current: "expand" | "collapse" = "collapse";
    let animation = null as Animation | null;

    onMounted(() => {
        if (initialized) {
            return;
        }

        if ("matchMedia" in window) {
            const prefersReducedMotion = window.matchMedia(
                "(prefers-reduced-motion: reduce)",
            );

            reducedMotion.value = prefersReducedMotion.matches;

            prefersReducedMotion.addEventListener("change", (event) => {
                reducedMotion.value = event.matches;
            });
        } else {
            /* fallback when matchMedia is not available, typically in jsdom under testing or similar environment */
            reducedMotion.value = true;
        }

        initialized = true;
    });

    watchEffect(() => {
        if (elementRef.value) {
            elementRef.value.classList.toggle("expanded", current === "expand");
        }
    });

    return {
        enabled: computed(() => reducedMotion.value === false),
        animate(state) {
            current = state;

            const element = elementRef.value;
            if (!element) {
                return;
            }

            element.classList.toggle("expanded", state === "expand");

            if (reducedMotion.value) {
                return;
            }

            if (animation) {
                animation.cancel();
            }

            element.classList.add("animating");
            const h = element.offsetHeight;

            if (state === "expand") {
                animation = element.animate(
                    [{ height: 0 }, { height: `${String(h)}px` }],
                    {
                        duration,
                        easing,
                    },
                );
            } else {
                animation = element.animate(
                    [{ height: `${String(h)}px` }, { height: 0 }],
                    {
                        duration,
                        easing,
                    },
                );
            }

            animation.addEventListener("finish", () => {
                element.classList.remove("animating");
            });
        },
    };
}
