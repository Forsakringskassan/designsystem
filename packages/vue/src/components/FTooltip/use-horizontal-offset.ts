import { onMounted, onUnmounted, readonly, type Ref, ref, watch } from "vue";

/**
 * Compute the offset from an alement to an ancestor container.
 *
 * @internal
 */
export function useHorizontalOffset(options: {
    element: Ref<HTMLElement | null>;
    parent: Ref<HTMLElement | null>;
}): Readonly<Ref<number>> {
    const { element: elementRef, parent: parentRef } = options;
    const offset = ref(0);

    watch(() => elementRef.value, updateOffset);
    watch(() => parentRef, updateOffset);

    /* @todo debounce */
    onMounted(() => window.addEventListener("resize", updateOffset));
    onUnmounted(() => window.removeEventListener("resize", updateOffset));

    return readonly(offset);

    function updateOffset(): void {
        const element = elementRef.value;
        const parent = parentRef.value;
        if (!element || !parent) {
            return;
        }

        /*
         * Container
         * +---------------------------+
         * |         [xxxx] - Element  |
         * +---------------------------+
         *            ^^
         *              \------ center (half the width of the element)
         * ^^^^^^^^^^
         *           \--------- left (distance between container left edge and
         *                            element left edge)
         *
         * The final offset is the left distance + half the width of the
         * element, adjusted for the border width of the arrow.
         */

        /* defer one tick to allow rendering to settle */
        setTimeout(() => {
            const borderWidth = 2;
            const center = element.offsetWidth / 2;
            const left = element.offsetLeft - parent.offsetLeft;
            offset.value = left - borderWidth + center;
        }, 0);
    }
}
