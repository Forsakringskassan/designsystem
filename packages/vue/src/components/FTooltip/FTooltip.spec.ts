import { VueWrapper, mount } from "@vue/test-utils";
import {
    cjsResolver,
    FileSystemConfigLoader,
    HtmlValidate,
} from "html-validate/node";
import "html-validate/jest";
import FTooltip from "./FTooltip.vue";

const tooltipButtonClass = ".tooltip__button";
const headerClass = ".tooltip__header";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FTooltip, {
        attrs: { ...attrs },
        props: { screenReaderText: "Lorem ipsum", ...props },
        slots: { ...slots },
        global: {
            stubs: ["f-icon"],
        },
    });
}

it("should set aria-expanded on tooltip button", () => {
    expect.assertions(2);
    const expanded = mount(FTooltip, {
        props: { screenReaderText: "", modelValue: true },
    });
    const collapsed = mount(FTooltip, {
        props: { screenReaderText: "", modelValue: false },
    });
    const selector = ".tooltip__button";
    expect(expanded.get(selector).attributes("aria-expanded")).toBe("true");
    expect(collapsed.get(selector).attributes("aria-expanded")).toBe("false");
});

it("should set expanded class on tooltip container", () => {
    expect.assertions(2);
    const expanded = mount(FTooltip, {
        props: { screenReaderText: "", modelValue: true },
    });
    const collapsed = mount(FTooltip, {
        props: { screenReaderText: "", modelValue: false },
    });
    const selector = ".tooltip";
    expect(expanded.get(selector).classes()).toContain("expanded");
    expect(collapsed.get(selector).classes()).not.toContain("expanded");
});

describe("slots", () => {
    it("should render header if header slot is used", async () => {
        const wrapper = createWrapper({
            slots: {
                header: `Tooltip`,
            },
            props: {
                headerTag: "h3",
            },
        });
        await wrapper.get(tooltipButtonClass).trigger("click");
        const header = wrapper.get(headerClass);
        expect(header.text()).toBe("Tooltip");
    });

    it("should throw error if has slot for header but no header-tag", async () => {
        expect.assertions(1);
        expect(() => {
            createWrapper({
                slots: {
                    header: `Tooltip`,
                },
            });
        }).toThrowErrorMatchingInlineSnapshot(
            `"Tooltip with header must define headerTag"`,
        );
    });

    it("should not render header if header slot isn't used", async () => {
        const wrapper = createWrapper();

        await wrapper.get(tooltipButtonClass).trigger("click");
        const header = wrapper.find(headerClass);
        expect(header.exists()).toBeFalsy();
    });
});

describe("html-validate", () => {
    const loader = new FileSystemConfigLoader([cjsResolver()], {
        extends: [
            "html-validate:recommended",
            "html-validate-vue:recommended",
            "@fkui/vue:recommended",
        ],
        plugins: ["<rootDir>/htmlvalidate", "html-validate-vue"],
    });
    const htmlvalidate = new HtmlValidate(loader);

    it("close-button-text", () => {
        expect.assertions(1);
        const markup = /* HTML */ `
            <!-- should allow missing close-button-text -->
            <f-tooltip screen-reader-text="foo"></f-tooltip>

            <!-- should not allow omitted close-button-text value -->
            <f-tooltip close-button-text screen-reader-text="foo"></f-tooltip>

            <!-- should not allow empty close-button-text value -->
            <f-tooltip
                close-button-text=""
                screen-reader-text="foo"
            ></f-tooltip>

            <!-- should allow valid close-button-text value -->
            <f-tooltip
                close-button-text="lorem ipsum"
                screen-reader-text="foo"
            ></f-tooltip>
        `;
        const report = htmlvalidate.validateString(markup);
        expect(report).toMatchInlineCodeframe(`
            "error: Attribute "close-button-text" is missing value (attribute-allowed-values) at inline:6:24:
              4 |
              5 |             <!-- should not allow omitted close-button-text value -->
            > 6 |             <f-tooltip close-button-text screen-reader-text="foo"></f-tooltip>
                |                        ^^^^^^^^^^^^^^^^^
              7 |
              8 |             <!-- should not allow empty close-button-text value -->
              9 |             <f-tooltip
            Selector: f-tooltip:nth-child(2)
            error: Attribute "close-button-text" has invalid value "" (attribute-allowed-values) at inline:10:17:
               8 |             <!-- should not allow empty close-button-text value -->
               9 |             <f-tooltip
            > 10 |                 close-button-text=""
                 |                 ^^^^^^^^^^^^^^^^^
              11 |                 screen-reader-text="foo"
              12 |             ></f-tooltip>
              13 |
            Selector: f-tooltip:nth-child(3)"
        `);
    });

    it("header-tag", () => {
        expect.assertions(1);
        const markup = /* HTML */ `
            <!-- should allow missing header-tag -->
            <f-tooltip screen-reader-text="foo"></f-tooltip>

            <!-- omitted header-tag value -->
            <f-tooltip header-tag screen-reader-text="foo"></f-tooltip>

            <!-- empty header-tag -->
            <f-tooltip header-tag="" screen-reader-text="foo"></f-tooltip>

            <!-- valid header-tag -->
            <f-tooltip header-tag="h1" screen-reader-text="foo"></f-tooltip>
            <f-tooltip header-tag="h2" screen-reader-text="foo"></f-tooltip>
            <f-tooltip header-tag="h3" screen-reader-text="foo"></f-tooltip>
            <f-tooltip header-tag="h4" screen-reader-text="foo"></f-tooltip>
            <f-tooltip header-tag="h5" screen-reader-text="foo"></f-tooltip>
            <f-tooltip header-tag="h6" screen-reader-text="foo"></f-tooltip>

            <!-- invalid header-tag -->
            <f-tooltip header-tag="foobar" screen-reader-text="foo"></f-tooltip>
            <f-tooltip header-tag="div" screen-reader-text="foo"></f-tooltip>
            <f-tooltip header-tag="span" screen-reader-text="foo"></f-tooltip>
        `;
        const report = htmlvalidate.validateString(markup);
        expect(report).toMatchInlineCodeframe(`
            "error: Attribute "header-tag" is missing value (attribute-allowed-values) at inline:6:24:
              4 |
              5 |             <!-- omitted header-tag value -->
            > 6 |             <f-tooltip header-tag screen-reader-text="foo"></f-tooltip>
                |                        ^^^^^^^^^^
              7 |
              8 |             <!-- empty header-tag -->
              9 |             <f-tooltip header-tag="" screen-reader-text="foo"></f-tooltip>
            Selector: f-tooltip:nth-child(2)
            error: Attribute "header-tag" has invalid value "" (attribute-allowed-values) at inline:9:24:
               7 |
               8 |             <!-- empty header-tag -->
            >  9 |             <f-tooltip header-tag="" screen-reader-text="foo"></f-tooltip>
                 |                        ^^^^^^^^^^
              10 |
              11 |             <!-- valid header-tag -->
              12 |             <f-tooltip header-tag="h1" screen-reader-text="foo"></f-tooltip>
            Selector: f-tooltip:nth-child(3)
            error: Attribute "header-tag" has invalid value "foobar" (attribute-allowed-values) at inline:20:36:
              18 |
              19 |             <!-- invalid header-tag -->
            > 20 |             <f-tooltip header-tag="foobar" screen-reader-text="foo"></f-tooltip>
                 |                                    ^^^^^^
              21 |             <f-tooltip header-tag="div" screen-reader-text="foo"></f-tooltip>
              22 |             <f-tooltip header-tag="span" screen-reader-text="foo"></f-tooltip>
              23 |
            Selector: f-tooltip:nth-child(10)
            error: Attribute "header-tag" has invalid value "div" (attribute-allowed-values) at inline:21:36:
              19 |             <!-- invalid header-tag -->
              20 |             <f-tooltip header-tag="foobar" screen-reader-text="foo"></f-tooltip>
            > 21 |             <f-tooltip header-tag="div" screen-reader-text="foo"></f-tooltip>
                 |                                    ^^^
              22 |             <f-tooltip header-tag="span" screen-reader-text="foo"></f-tooltip>
              23 |
            Selector: f-tooltip:nth-child(11)
            error: Attribute "header-tag" has invalid value "span" (attribute-allowed-values) at inline:22:36:
              20 |             <f-tooltip header-tag="foobar" screen-reader-text="foo"></f-tooltip>
              21 |             <f-tooltip header-tag="div" screen-reader-text="foo"></f-tooltip>
            > 22 |             <f-tooltip header-tag="span" screen-reader-text="foo"></f-tooltip>
                 |                                    ^^^^
              23 |
            Selector: f-tooltip:nth-child(12)"
        `);
    });

    it("screen-reader-text", () => {
        expect.assertions(1);
        const markup = /* HTML */ `
            <!-- should not allow missing screen-reader-text -->
            <f-tooltip></f-tooltip>

            <!-- should not allow omitted screen-reader-text value -->
            <f-tooltip screen-reader-text></f-tooltip>

            <!-- should not allow empty screen-reader-text value -->
            <f-tooltip screen-reader-text=""></f-tooltip>

            <!-- should allow valid screen-reader-text -->
            <f-tooltip screen-reader-text="lorem ipsum"></f-tooltip>
        `;
        const report = htmlvalidate.validateString(markup);
        expect(report).toMatchInlineCodeframe(`
            "error: <f-tooltip> is missing required "screen-reader-text" attribute (element-required-attributes) at inline:3:14:
              1 |
              2 |             <!-- should not allow missing screen-reader-text -->
            > 3 |             <f-tooltip></f-tooltip>
                |              ^^^^^^^^^
              4 |
              5 |             <!-- should not allow omitted screen-reader-text value -->
              6 |             <f-tooltip screen-reader-text></f-tooltip>
            Selector: f-tooltip:nth-child(1)
            error: Attribute "screen-reader-text" is missing value (attribute-allowed-values) at inline:6:24:
              4 |
              5 |             <!-- should not allow omitted screen-reader-text value -->
            > 6 |             <f-tooltip screen-reader-text></f-tooltip>
                |                        ^^^^^^^^^^^^^^^^^^
              7 |
              8 |             <!-- should not allow empty screen-reader-text value -->
              9 |             <f-tooltip screen-reader-text=""></f-tooltip>
            Selector: f-tooltip:nth-child(2)
            error: Attribute "screen-reader-text" has invalid value "" (attribute-allowed-values) at inline:9:24:
               7 |
               8 |             <!-- should not allow empty screen-reader-text value -->
            >  9 |             <f-tooltip screen-reader-text=""></f-tooltip>
                 |                        ^^^^^^^^^^^^^^^^^^
              10 |
              11 |             <!-- should allow valid screen-reader-text -->
              12 |             <f-tooltip screen-reader-text="lorem ipsum"></f-tooltip>
            Selector: f-tooltip:nth-child(3)"
        `);
    });
});
