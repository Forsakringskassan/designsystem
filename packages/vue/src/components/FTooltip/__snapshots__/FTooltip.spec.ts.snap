// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with header and content when open 1`] = `
<div
  class="tooltip"
>
  <div
    class="tooltip__container"
  >
    <button
      aria-expanded="true"
      class="tooltip__button"
      type="button"
    >
      <span
        class="icon-stack icon-stack--tooltip"
      >
        <f-icon-stub
          library="f"
          name="circle"
        />
        <f-icon-stub
          library="f"
          name="i"
        />
        <span
          class="sr-only"
        >
          Screen reader text
        </span>
      </span>
    </button>
    <transition-stub
      appear="false"
      css="true"
      persisted="false"
    >
      
      <div>
        <div
          class="tooltip__content-wrapper"
          tabindex="-1"
        >
          <span
            class="tooltip__arrow"
            style="left: -2px;"
          />
          <div
            class="tooltip__content"
          >
            <div
              class="tooltip__header"
            >
              
              Tooltip
              
            </div>
            <div
              class="tooltip__body"
            >
              
              <p>
                Content...
              </p>
              
            </div>
          </div>
          <div
            class="iflex iflex--float-right"
          >
            
            <div
              class="iflex__item iflex--align-top iflex--shrink"
            >
              
              <button
                class="close-button"
                type="button"
              >
                <span>
                  Close
                </span>
                <f-icon-stub
                  class="button__icon"
                  library="f"
                  name="close"
                />
              </button>
              
            </div>
            
          </div>
        </div>
      </div>
      
    </transition-stub>
  </div>
</div>
`;

exports[`snapshots should match snapshot with header, content and screen reader text 1`] = `
<div
  class="tooltip"
>
  <div
    class="tooltip__container"
  >
    <button
      aria-expanded="false"
      class="tooltip__button"
      type="button"
    >
      <span
        class="icon-stack icon-stack--tooltip"
      >
        <f-icon-stub
          library="f"
          name="circle"
        />
        <f-icon-stub
          library="f"
          name="i"
        />
        <span
          class="sr-only"
        >
          Screen reader text
        </span>
      </span>
    </button>
    <transition-stub
      appear="false"
      css="true"
      persisted="false"
    >
      
      
    </transition-stub>
  </div>
</div>
`;
